### ISSUE TITLE
syntax: "indent" grouping

### ISSUE THREAD

**Issue by raichoo** [neovim/neovim#6052]
There are several popular languages (Haskell, Python) and file formats (YAML, cabal) out there that are indentation sensitive and as far as I know `Vim` is not very good in supporting that when it comes to things like syntax highlighting.
The only approach that I'm aware of is doing backtracking with Vimscript which is tedious and error prone.
I'm the main author of `haskell-vim` and I get quite a lot of feature requests that would need a way to figure out in which kind of block I am.

A very handy thing would be something like `syn indent` that works like so:

Let's say we want to define how `do` blocks in Haskell are being highlighted. They start with the keyword `do`.

```viml
syn indent {group-name} start="\<do\>" contains=…
```
If that rule finds a match all the following lines indented relative to the match belong to that group.

Example:

```haskell
fun x = do -- match
  foo -- indented by 2 relative to the matched line, therefore part of the group
  bar -- ditto

-- not indented anymore, therefore not part of the group. 
bar x d = …
```

A way to tell that the group should ignore empty lines would also be handy so something like this could work:

```haskell
fun x = do --match
   foo -- part of the group

   bar -- also part of the group
```
Maybe there already is a way to do this with Neovim but if there isn't this would make writing plugins  for languages with significant whitespace a lot easier.

**Comment by justinmk** [comment]
Indentation as a first-class concept in Vim would be useful for many things, e.g. navigation and an "indent" text-object.

@ZyX-I Is the maintainer of YAML indent plugin, maybe he can comment on whether the current capabilities are sufficient.

**Comment by ZyX-I** [comment]
@justinmk I have never touched `:syn` and did not write many syntax scripts, but AFAIK properly supporting indentation-based languages is impossible. You may see that my YAML *syntax* script does not even attempt to do something like this. Indentation scripts are irrelevant.

// And I would really like to see syntax based on calling functions rather then extending regex-based `:syn`. If properly implemented this should be both faster (needs to allow something that touches VimL least: maybe, new API function + #4971 + API event regarding changing text, #4971 is needed to avoid unnecessary allocations that should slow displaying syntax the most) and allow supporting *any* syntax correctly. And allow adding features like “highlighting C functions as functions and C macros as macros” without trying to jump above one’s head to properly implement this using `:syn` and VimL.

**Comment by raichoo** [comment]
I agree, indentation scripts are really painful to write, but I'm not sure what you are proposing here and how one would write indentation for a specific language for neovim using the approach you just described. Could you elaborate on that a little more? Having better control over indentation than one currently has with `:syn` would be really nice.

**Comment by ZyX-I** [comment]
@raichoo Vim has no problems with writing indentation scripts, there is nothing to improve here. And `:syn` has no control over indentation, so I do not understand what you are asking. I did not propose anything for “writing indentation”.

**Comment by raichoo** [comment]
Ah sorry I didn't not make myself perfectly clear here. Yes, of course `:syn` is used to define syntax (or whatever the correct Vim terminology is here). That alone would help me a lot in terms of highlighting, let's say to allow different highlighting in different kinds of blocks (`let`, `do`, `where` just to name a few Haskell examples), I can already do that with regions but it's, as far as I know, impossible to do that with indented blocks.
Also: you can use that syntax information to gain more information during indentation using `synIDAttr`. I do that quite a lot in `haskell-vim` because different contexts may need different indentation behavior.

I disagree with the statement that Vim has no problems when it comes to indentation. Nearly everyone I know who attempted to write a more sophisticated indentation Script in Vim gave up at some point because Vim features don't provide the needed features and reimplementig the needed features in VimL for every new plugin is a huge pain.
I regularly have to turn close issues on `haskell-vim` by people asking me to add a different kind of indentation behavior to `do` or `let` blocks because I have not tools other than manual backtracking to find out if I'm in one of these blocks, and as I said before: that's tedious to implement, error-prone and slow. Adding a `:syn indent` feature or anything akin to that would me allow to use `synIDAttr` to figure out if I'm in such a block the same way I can use it right now to figure out if I'm in a region.

**Comment by ZyX-I** [comment]
@raichoo About indentation: I actually remembered that there `&indentexpr` does not allow smart tabs (indent with tabs, align with spaces), but this is the only thing I can remember which is missing and I usually do not use tabs.

I now understand what’s your problem with `:syn` and why @justinmk was mentioning indentation scripts, but I would want something more generic: allow syntax to be defined by functions which would return something like `[[[1, 1], [1, 7], 3], [[1, 6], [1, 7], 4], …]` (highlight first .. seventh byte of the first line with highlight group 3, sixth to seventh byte of the first line with highlight group 4, …): this should allow arbitrary highlighting which may be based also on such properties as “this is a file-local (static) function”, “this is an external function”, “this is a macro call”. Exact API does not need to be exactly the same as described, but it must be fast (this is why I used numbers for highlight groups and not more readable strings); better first check how IDEs cope with such things. “Arbitrary highlighting” like described here means that highlighting may and should be implemented based on syntax parsers actually used by compilers (if they export this somehow, of course) as opposed to trying to do syntax parsing based on rewriting parsers in regexes which do not cope well with some grammars. AFAIR there was an issue for this somewhere.

**Comment by raichoo** [comment]
What you describe sounds like it could also enable things like semantic highlighting which is pretty much the "holy grail" I've been looking for for quite a while. Yes, reimplementing the parsers with regexes simply has too many limitations and I believe most plugin writers struggle with that a lot (at least I do).
I'm not quite sure how using the parser of the compiler (if exposed at all) might help here though since most of them require a syntax correct file and don't really allow partially parsing a file to extract syntax information. So I'm not quite sure how the highlighting information is obtained other then reimplementing the parser yet again (in something that is hopefully more expressive and faster than VimL ^^) and maybe pulling tricks like building the derivative of the languages being parsed (to avoid parsing the whole file over and over again, when only a few bits at the and changed). But I that's another issue, I'm just trying to figure out where this might fit into the bigger picture.
Anyway, I'm generally intrigued by the idea to have such fine grained control over highlighting and all the good things that come with that. For a lot of tasks however it sounds a bit like "overkill" when quickly hacking together something with `:syn` and regex might do the job. My proposal is aimed at being able to quickly write such a highlighter. For `haskell-vim` however I would certainly switch to an API that gives me the finer control if it is not too much of a pain to get the job done.

**Comment by ZyX-I** [comment]
@raichoo Some compilers are able to proceed after encountering syntax errors. And sometimes it is not impossible to push patches for such purposes (though I guess usually it is easier to do reimplementing as I have not heard about such initiatives). And there are languages where it is completely possible to use the strategy “take the grammar files from project and make use of them for constructing your syntax parser”.

In any case, even if you can’t reuse compiler’s parser directly for syntax highlighting you most likely will have to use it indirectly for semantic highlighting.

**Comment by justinmk** [comment]
`nvim_buf_add_highlight()` was added for this purpose. What is lacking from it?

**Comment by raichoo** [comment]
I've played with `nvim_buf_add_highlight()` and even though it keeps the highlighting intact when lines are moved around, the same does not hold for column changes which IMHO makes it not very useful for writing highlighters for constantly changing files.

![screenshot from 2017-02-04 09-58-07](https://cloud.githubusercontent.com/assets/151506/22617272/df09b910-eac0-11e6-9f9c-ea272bbfbe32.png)

![screenshot from 2017-02-04 09-58-29](https://cloud.githubusercontent.com/assets/151506/22617273/e237c870-eac0-11e6-8f0c-414c39138484.png)

The `nvim_buf_*` operations all seem a bit too low level for my taste when it comes to writing rather highlevel plugins like highlighters.

**Comment by bounceme** [comment]
https://github.com/neovim/neovim/issues/6052#issuecomment-277404351

does it extend to the `synid()` type functions? that is one of the most powerful tools in indent scripts

https://github.com/pangloss/vim-javascript/blob/master/indent/javascript.vim#L224 is what i used for braceless javascript statements. the idea is to find how many 'levels' are needed to add indent for. 

```
if (1)
  if (2)
    if (3)
      return 1 -
        2
let a = b
```

**Comment by bounceme** [comment]
oh. this about syntax based on indentation, not indenting. carry on. hah

**Comment by bounceme** [comment]
@raichoo you can use `synstack()` as long as the syntax knows the start and end of each block.

**Comment by tweekmonster** [comment]
Indented blocks is something I tried to tackle with [braceless.vim](https://github.com/tweekmonster/braceless.vim).  It works pretty well but is a little heavy-handed with regex.  Since it was my first "serious" plugin, I went a little overboard with its features and it needs to be refactored.  But, I tried to make it generic enough to serve as a kind of framework for indented filetypes.

Every time I think about refactoring it, I come back to the thought that it would be best if Vim had a builtin PEG/BNF parser so it could actually understand the structure of certain filetypes.  I guess Haskell doesn't have such a grammar description, but in [Python's][1], indented bodies could be handled by finding `suite` symbols.  The problem is I have absolutely no idea how to even get started with it 😬

[1]: https://github.com/python/cpython/blob/master/Grammar/Grammar

**Comment by justinmk** [comment]
@tweekmonster [scintillua](https://foicica.com/scintillua/) has a bunch of PEG lexers already written. Even [vis editor](https://github.com/martanne/vis/tree/master/lua) uses them. Lua will be built-in to Nvim before the end of the month...

They are lexers, not parsers. But it's using LPEG so there's already something to work with.

**Comment by tweekmonster** [comment]
@justinmk That's pretty awesome!  I'll need to spruce up my Lua to tinker with it.

The lexers would still be good for simplifying syntax highlighting.  scintillua could make it possible to label text regions the way TextMate does, too.  Somewhat related article about how VSCode stores grammar tokens: https://code.visualstudio.com/blogs/2017/02/08/syntax-highlighting-optimizations

**Comment by bounceme** [comment]
If plugins switch to the new api for highlighting that will affect many plugins that use syntax information. Autocompletion , indent , tons of scripts in default runtime, etc.

**Comment by raichoo** [comment]
@bounceme That's the whole point of my request, knowing where the block starts and where it ends. I'm already using `synstack()` in `haskell-vim` but I can't to this for blocks that are delimited by their indentation.

@tweekmonster: There is a syntax reference for Haskell https://www.haskell.org/onlinereport/syntax-iso.html.

**Comment by tweekmonster** [comment]
> If plugins switch to the new api for highlighting that will affect many plugins that use syntax information. Autocompletion , indent , tons of scripts in default runtime, etc.

@bounceme Are you talking about my comment about simplifying syntax highlighting?  I wouldn't think so.  In my mind, it would be something like:

```vim
syntax lex pythonFunction python.function
```

Besides, syntax-dependent matching has always been flaky in my experience.  Syntax group names aren't exactly codified and using a plugin for improved syntax will eventually break some plugin that only expected the built-in groups.

> There is a syntax reference for Haskell 

I stand corrected!  Is there a loadable text version of it that's maintained?

---

I've been mulling over this issue a little.  Expanding on the example @raichoo first posted, It might be nice to define `syn-indent` as something like:

```vim
syntax region indentDoBlock start="\<do\>" skip="^\s*$" transparent indent=1
```

Except `end=` is ignored since it would be inferred.  The end is found by scanning the subsequent lines for indentation that's >= `indent() + synindent`, while skipping lines that shouldn't be evaluated, and stopping at a line that has printable text and indentation < `indent() + synindent`.  It's probably not the best idea, but a companion function `synregion(group[, lnum, col])` could return the start and end of syntax group for scripting.  It'd make it easier to have operators that are based on the how the user sees the syntax.  Like `ci"` could work the way you'd expect for multiline strings.

@ZyX-I The biggest problem I had with `&indentexpr` (and `&foldexpr`) for indented languages is that there's no way to tell them to skip lines.  When I was working on indent/folding for Python, each line that's evaluated required scanning backwards to determine the correct indentation level for the current line.  It really slowed things down if a function body spanned 100 lines when using the `=` motion.  After finding the start of the indented block, it would be possible to scan forward to get the end of the indent block and return the indent levels of all lines involved in that single `&indentexpr` call.  In a Python script that has 5000 lines and no indentation, `&indentexpr` could only need one call.  The best way to deal with this currently is to cache the work for subsequent `&indentexpr` calls, but it still results in wasted time and requires cache management.

**Comment by bounceme** [comment]
I meant for `nvim_buf_add_highlight()` . my issue is that even if highlighting is semantically perfect using the `Matchaddpos` like function, no information will be provided to other vimscript which uses the conventional `synId()` and `synstack()` functions

**Comment by bounceme** [comment]
> @ZyX-I The biggest problem I had with &indentexpr (and &foldexpr) for indented languages is that there's no way to tell them to skip lines.

`prevnonblank()` ? or :
```
if getline('.') is ''
  return -1
endif
````

**Comment by tweekmonster** [comment]
> prevnonblank() ? or :

No, I'm not talking about skipping blank lines.  If the `&indentexpr` function is run on an entire script like this:

```python
   1. def test():
   2.     # body start
3-99.     # ...
 100.     # body end
```

Since I can parse the function body as an indented block for operators, I can determine what the indent level is for lines 2-100 because of what's on line 1, but I can only return the indent level for the current line being evaluated.  `&indentexpr` is run on each line, and in each call the block is parsed scanning back to line 1 to determine the current line's indent level.  So, work is repeated on each line and taking longer as the line number increases.  Additionally, caching is a pain because there isn't an easy way to create a cache that's only for the current `&indentexpr` range.

This isn't a huge problem when it's called while editing with `&smartindent` enabled, but reformatting a large range becomes painfully slow.  It's the same problem with `&foldexpr`, but it leads to a worse experience since it's ran on an entire buffer when it's loaded.

**Comment by bounceme** [comment]
Yea, random access lexing would be very hard for that example, https://github.com/pangloss/vim-javascript/blob/master/indent/javascript.vim#L224 was how I did something similar. that works for when only one statement is expected in a braceless context though

**Comment by justinmk** [comment]
This is covered by the generalized textobjects topic: 

- https://github.com/neovim/neovim/issues/8544
